core/src/avm1/globals/system.rs

```diff
@@ -290,32 +339,397 @@ pub struct SystemProperties {
+            _ => Language::Unknown,
+        };
+
+        let language_region = if language == Language::English && player_version < 7 {
+            if language_vector.len() > 1 {
+                Some(language_vector[1].to_uppercase())
+            } else {
+                Some("US".to_string())
+            }
+        } else {
+            None
+        };
+
+        #[cfg(not(target_family = "wasm"))]
+        fn get_os() -> OperatingSystem {
+            pub const OS: &str = std::env::consts::OS;
+
+            return match OS {
+                "linux" | "android" | "freebsd" | "dragonfly" | "netbsd" | "openbsd"
+                | "solaris" => OperatingSystem::Linux,
+                "macos" | "ios" => OperatingSystem::MacOs,
+                "windows" => {
+                    #[allow(clippy::zero_prefixed_literal)]
+                    fn match_windows_version(
+                        major_version: u64,
+                        minor_version: u64,
+                        server_version: bool,
+                    ) -> OperatingSystem {
+                        let version = (major_version, minor_version, server_version);
+                        // Not including all Windows versions since Windows versions
+                        // Ruffle can't run on don't need to be detected
+                        match version {
+                            (10, 0, _) | (6, 4, _) | (6, 3, _) | (6, 2, _) => {
+                                OperatingSystem::Windows8
+                            }
+                            (6, 1, true) => OperatingSystem::WindowsServer2008R2,
+                            (6, 1, false) => OperatingSystem::Windows7,
+                            (6, 0, true) => OperatingSystem::WindowsServer2008,
+                            (6, 0, false) => OperatingSystem::WindowsVista,
+                            (5, 2, true) => OperatingSystem::WindowsServer2003R2,
+                            _ => OperatingSystem::WindowsUnknown,
+                        }
+                    }
+
+                    if let os_info::Version::Semantic(major_version, minor_version, _) =
+                        *os_info::get().version()
+                    {
+                        let server_version = os_info::get()
+                            .edition()
+                            .unwrap_or("")
+                            .to_lowercase()
+                            .contains("server");
+                        match_windows_version(major_version, minor_version, server_version)
+                    } else {
+                        OperatingSystem::WindowsUnknown
+                    }
+                }
+
+                // Fallback
+                _ => OperatingSystem::Linux,
+            };
+        }
+
+        #[cfg(target_family = "wasm")]
+        fn get_os() -> OperatingSystem {
+            let window = web_sys::window().unwrap();
+            let user_agent = window.navigator().user_agent().unwrap();
```

Comment:
Would getting the user agent/operating system be better moved to an implementation on the navigator backend?

Reply:
Yeah, I see that point (although `ExternalNavigatorBackend`, `TestNavigatorBackend` and `NullNavigatorBackend` would share the same implementation (but this could easily be done with a function all three call)).
Do you prefer this? Then I can change it like that :)
@Aaron1011 (I hope the ping is okay) what do you think about this?

Reply:
I don't have a strong preference either way.